File: LR2/27/27.php
<?php
// Чтение всей строки из стандартного ввода
$input = trim(fgets(STDIN));

// Разделение входной строки на массив чисел по пробелам
$tokens = explode(' ', $input);

// Первый элемент - количество чисел
$nums = intval($tokens[0]);

for ($i = 1; $i <= $nums; $i++) {
    $token = $tokens[$i];
    $validInput = true;
    
    // Проверка, что все символы - цифры
    for ($j = 0; $j < strlen($token); $j++) {
        $symbol = ord($token[$j]) - ord('0');
        if ($symbol < 0 || $symbol > 9) {
            $validInput = false;
            break;
        }
    }
    
    // Если входная строка не прошла проверку или содержит менее двух цифр
    if (!$validInput || strlen($token) < 2) {
        echo "NO";
        continue;
    }
    
    $isCorrectDate = false;
    $number = intval($token);
    
    if ($number >= 11 && $number < 100) {
        $day = intval($number / 10);
        $month = $number % 10;
        $isCorrectDate = ($day >= 1 && $day <= 31 && $month >= 1 && $month <= 12);
    } else if ($number >= 100 && $number < 1000) {
        $day = intval($number / 10);
        $month = $number % 10;
        $isCorrectDate = ($day >= 1 && $day <= 31 && $month >= 1 && $month <= 12);
        if (!$isCorrectDate) {
            $day = intval($number / 100);
            $month = $number % 100;
            $isCorrectDate = ($day >= 1 && $day <= 31 && $month >= 1 && $month <= 12);
        }
    } else if ($number >= 1000 && $number <= 3112) {
        $day = intval($number / 100);
        $month = $number % 100;
        $isCorrectDate = ($day >= 1 && $day <= 31 && $month >= 1 && $month <= 12);
    }
    
    echo $isCorrectDate ? "YES" : "NO";
}
?>

File: LR2/27/27.rs
use std::io::{self, BufRead};

/// Проверяет, является ли дата корректной
fn is_correct_date(day: u32, month: u32) -> bool {
    day >= 1 && day <= 31 && month >= 1 && month <= 12
}

fn main() {
    let stdin = io::stdin();
    let mut input = String::new();
    stdin.lock().read_line(&mut input).expect("Ошибка при чтении строки");
    
    let tokens: Vec<&str> = input.trim().split_whitespace().collect();
    let nums: usize = tokens[0].parse().expect("Ошибка при парсинге количества чисел");
    
    for i in 1..=nums {
        if i >= tokens.len() {
            break;
        }
        
        let token = tokens[i];
        let mut valid_input = true;
        
        // Проверка, что все символы - цифры
        for c in token.chars() {
            if !c.is_ascii_digit() {
                valid_input = false;
                break;
            }
        }
        
        if !valid_input || token.len() < 2 {
            print!("NO");
            continue;
        }
        
        let number: u32 = token.parse().unwrap();
        let mut is_valid_date = false;
        
        if number >= 11 && number < 100 {
            let day = number / 10;
            let month = number % 10;
            is_valid_date = is_correct_date(day, month);
        } else if number >= 100 && number < 1000 {
            let day = number / 10;
            let month = number % 10;
            is_valid_date = is_correct_date(day, month);
            
            if !is_valid_date {
                let day = number / 100;
                let month = number % 100;
                is_valid_date = is_correct_date(day, month);
            }
        } else if number >= 1000 && number <= 3112 {
            let day = number / 100;
            let month = number % 100;
            is_valid_date = is_correct_date(day, month);
        }
        
        print!("{}", if is_valid_date { "YES" } else { "NO" });
    }
}